finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
-    // `best_block` is a `descendant` of the previous query, then we don't need to update proposed_block, as it is
-    // already correct.
-    let mut prev_best_hash: B::Hash = proposed_block.lock().hash;
-    let mut prev_best_number: NumberFor<B> = proposed_block.lock().number;
-    loop {
-        let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
-        tokio::select! {
-            _ = delay => {
-                let new_best_header = select_chain
-                    .best_chain()
-                    .await
-                    .expect("No best chain");
-                if *new_best_header.number() <= max_block_num {
-                    *proposed_block.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
-                } else {
-                    // we check if prev_best_header is an ancestor of new_best_header:
-                    if proposed_block.lock().number < max_block_num {
-                        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
-                        *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                    } else {
-                        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), prev_best_number);
-                        if reduced_header.hash() != prev_best_hash {
-                            // the new best block is not a descendant of previous best
-                            let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
-                            *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                        }
-                    }
+    pub(crate) fn new(
+        select_chain: SC,
+        client: Arc<C>,
+        min_block_num: NumberFor<B>,
+        max_block_num: NumberFor<B>,
+        metrics: Option<Metrics<<B::Header as HeaderT>::Hash>>,
+    ) -> (Self, DataProvider<B>) {
+        let initial_data = AlephData::Empty;
+        let data_to_propose = Arc::new(Mutex::new(initial_data));
+        (
+            ChainTracker {
+                select_chain,
+                client,
+                data_to_propose: data_to_propose.clone(),
+                min_block_num,
+                max_block_num,
+                best_block_in_session: None,
+                prev_chain_info: None,
+                _phantom: PhantomData,
+            },
+            DataProvider {
+                data_to_propose,
+                metrics,
+            },
+        )
+    }
+
+    fn update_data(&mut self) {
+        // We read the best_block_in_session field and the highest_finalized block from the client and compute
+        // the corresponding `AlephData<B>` in `data_to_propose` fort AlephBFT. To not recompute this many
```

Comment:
Fixed type. 
It will fire roughly 80% of the time with current parameters and will save lots of unnecessary client queries. I think it's worth it quite a bit.